### PR TITLE
chore: v2 upgrade handler

### PR DESCRIPTION
**Pull request by omritoptix**
This PR attempts to perform an upgrade to the upcoming v2.
It also attempts to migrate ibc clients from dymint to tendermint. 

# PR Standards

## Opening a pull request should be able to meet the following requirements:

---

For Author:

- [ ]  Targeted PR against correct branch
- [ ]  included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Linked to Github issue with discussion and accepted design
- [ ]  Targets only one github issue
- [ ]  Wrote unit and integration tests
- [ ]  All CI checks have passed
- [ ]  Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code)

---

For Reviewer:

- [ ]  confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ]  Reviewers assigned
- [ ]  confirmed all author checklist items have been addressed

---;

After reviewer approval:

- [ ]  In case targets main branch, PR should be squashed and merged.
- [ ]  In case PR targets a release branch, PR should be rebased.

### CODE REVIEW

File: app/upgrades/v2/upgrade.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+func CreateUpgradeHandler(
+	mm *module.Manager,
+	configurator module.Configurator,
+	bankKeeper bankkeeper.Keeper,
+	clientKeeper clientkeeper.Keeper,
+	rollappKeeper rollappKeeper.Keeper,
+	cdc codec.BinaryCodec,
+
+) upgradetypes.UpgradeHandler {
+	return func(ctx sdk.Context, _ upgradetypes.Plan, vm module.VersionMap) (module.VersionMap, error) {
```
Comment (mtsitrin) [inline app/upgrades/v2/upgrade.go:31]: break this logic into smaller units
`migrateIBC`, `migrateRollapps`, etc..

File: app/upgrades/v2/upgrade.go [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+			return false
+		})
+
+		// Update the denom metadata for DYM token
+		bankKeeper.SetDenomMetaData(ctx, DYMTokenMetata)
+
+		// Update the params for rolllapp to init the new param enable_rollapps
+		rollappKeeper.SetParams(ctx, rollapptypes.DefaultParams())
+
+		// Iterate over all rollapps and set PermissionedAddresses to empty as the migration
```
Comment (mtsitrin) [inline app/upgrades/v2/upgrade.go:97]: why u need it?

File: app/app.go [hunk 3]
```diff
@@ -709,8 +710,8 @@ func New(
 		// this line is used by starport scaffolding # stargate/app/appModule
 
 		// Ethermint app modules
-		evm.NewAppModule(app.EvmKeeper, app.AccountKeeper, app.GetSubspace(evmtypes.ModuleName)),
-		feemarket.NewAppModule(app.FeeMarketKeeper, app.GetSubspace(feemarkettypes.ModuleName)),
+		evm.NewAppModule(app.EvmKeeper, app.AccountKeeper, app.GetSubspace(evmtypes.ModuleName).WithKeyTable(evmtypes.ParamKeyTable())),
```
Comment (mtsitrin) [inline app/app.go:713]: what for?

Comment (omritoptix) [reply app/app.go:713]: migration doesn't work without. it's missing the key table when trying to upgrade.

File: app/upgrades/v2/upgrade.go (hunk 2 above)
Comment (omritoptix) [reply app/upgrades/v2/upgrade.go:97]: it's missing the rollapps_enabled params and fails on it.

File: scripts/setup_local.sh [hunk 4]
```diff
@@ -89,6 +89,7 @@ set_EVM_params
 set_bank_denom_metadata
 set_epochs_params
 set_incentives_params
+set_misc_params
```
Comment (mtsitrin) [inline scripts/setup_local.sh:92]: already called above

File: app/upgrades/v2/upgrade.go (hunk 2 above)
Comment (mtsitrin) [reply app/upgrades/v2/upgrade.go:53]: I mean the change in the `PermissionedAddresses`
Isn't it was solved by the proto file update?

File: go.mod [hunk 5]
```diff
@@ -215,8 +215,10 @@ replace (
 )
 
 replace (
+	github.com/cosmos/ibc-go/v6 => github.com/dymensionxyz/ibc-go/v6 v6.0.0-rc0.0.20231123152216-df3f798086fa
 	// use dymension forks
-	github.com/evmos/ethermint => github.com/dymensionxyz/ethermint v0.22.0-dymension-v0.2
+	github.com/evmos/ethermint => github.com/dymensionxyz/ethermint v0.22.0-dymension-v0.3
 	github.com/osmosis-labs/osmosis/osmomath => github.com/dymensionxyz/osmosis/osmomath v0.0.4-dymension-v0.1
 	github.com/osmosis-labs/osmosis/v15 => github.com/dymensionxyz/osmosis/v15 v15.2.0-dymension-v0.3
+	github.com/tendermint/tm-db => github.com/dymensionxyz/tm-db v0.1.0-v0.6.8-tmdb
```
Comment (mtsitrin) [inline go.mod:223]: u sure u want to introduce the `NoCompression` change?
feels too harsh IMO

File: app/upgrades/v2/upgrade.go [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+		updateConsensusStates(ctx, clientKeeper, logger)
+
+		// Verify the client statuses are the same after the upgrade
+		verifyClientStatus(ctx, clientKeeper, clientStatuses, logger, cdc)
+
+		// Update the denom metadata for DYM token
+		bankKeeper.SetDenomMetaData(ctx, DYMTokenMetata)
+
+		// Update the params for rolllapp to init the new param enable_rollapps
+		rollappKeeper.SetParams(ctx, rollapptypes.DefaultParams())
```
[the remark this answers is not in the document]
Comment (omritoptix) [reply app/upgrades/v2/upgrade.go:51]: it's the same in froopyland and the default. both are 3.

File: app/upgrades/v2/upgrade.go (hunk 2 above)
Comment (omritoptix) [reply app/upgrades/v2/upgrade.go:53]: no..

**Review by mtsitrin — APPROVED** [review]

**Pull request merged**